File: 0218-the-skyline-problem/0218-the-skyline-problem.java
/*
 * 218. The Skyline Problem  (Hard)
 *
 * Approach: line-sweep over building edges with a max-heap of active heights.
 *   Encode each building as TWO events at x = left (start) and x = right (end).
 *   Use a sign trick on the height to distinguish start vs end events:
 *     start -> store -h (negative)
 *     end   -> store +h (positive)
 *   Then sort events by (x, height). The signed-height tiebreak gives us the
 *   ordering we need at duplicate x's:
 *     - At the SAME x, starts (negative h) sort BEFORE ends (positive h).
 *       Critical when two buildings touch: a new building starting at x must
 *       be active before we drop the one ending at x, otherwise we'd emit a
 *       phantom dip down to 0.
 *     - Two starts at same x: taller first (more negative h sorts first) —
 *       prevents two key-points at same x.
 *     - Two ends at same x: shorter first — prevents intermediate emission.
 *
 *   Maintain a max-heap of CURRENTLY ACTIVE heights (seeded with 0 for ground).
 *   On a start event, push h; on an end event, remove h. After each event,
 *   compare currentMax = pq.peek() to prevMax — if changed, emit (x, currentMax).
 *
 *      buildings (skyline outline = key points where height changes):
 *               ____
 *           ___|    |___
 *          |    ____|   |
 *      ____|   |        |__
 *
 *   Note: pq.remove(h) on Java PQ is O(n). For 218 a TreeMap<height,count>
 *   gives O(log n) and the same answer; this version is judge-accepted.
 *
 * Time: O(n^2) worst (PQ.remove)   Space: O(n)
 */
class Solution {
    public List<List<Integer>> getSkyline(int[][] buildings) {
       List<List<Integer>> result = new ArrayList<>();
       List<int[]> heights = new ArrayList<>();
       transform(buildings,heights);
       // sort by x, then by signed height: starts (negative) before ends (positive)
       Collections.sort(heights,(a,b)-> (a[0] == b[0])? a[1]-b[1] : a[0] - b[0]);
       // max-heap of active heights; seed with 0 = ground level
       PriorityQueue<Integer> pq = new PriorityQueue<Integer>((a,b)->(b-a));
       pq.offer(0);
       int prevMax =0;
       int currMax = 0;

       for (int[] height : heights){
           if(height[1] <0){ //start of building (h was stored negative)
                pq.offer(-height[1]);
           }
           else{ // end of building
                pq.remove(height[1]);
           }
           currMax = pq.peek();
           if(prevMax != currMax ){
                List<Integer> sub = new ArrayList<>(2);
                sub.add(height[0]);
                sub.add(currMax);
                result.add(sub);
                prevMax = currMax;
           }
       }

       return result;
    }

    private void transform(int[][] buildings,List<int[]> heights ){
            for(int[]build : buildings){
                heights.add(new int[]{build[0], -build[2]});
                heights.add(new int[]{build[1], build[2]});
            }

    }
}

File: 0218-the-skyline-problem/0218-the-skyline-problem.kt
import java.util.PriorityQueue

/*
 * 218. The Skyline Problem  (Hard)
 *
 * Approach: line-sweep over building edges with a max-heap of active heights.
 *   Encode each building as TWO events with a sign trick on height:
 *     start -> store -h     end -> store +h
 *   Sort events by (x, signedHeight). The tie-break is what makes this work:
 *     - At same x, starts (negative h) come BEFORE ends (positive h) -> two
 *       buildings that touch don't emit a phantom dip to 0.
 *     - Two starts at same x: taller first (more-negative sorts first) -> only
 *       one key-point emitted.
 *     - Two ends at same x: shorter first -> no intermediate emission.
 *
 *   Maintain a max-heap of currently active heights (seeded with 0). On start,
 *   push h; on end, remove h. After each event, if peek() differs from prevMax,
 *   emit (x, peek()).
 *
 *      Skyline = the set of x's where the max active height changes.
 *
 * Time: O(n^2) worst (heap.remove)   Space: O(n)
 */
class Solution {
    fun getSkyline(buildings: Array<IntArray>): List<List<Int>> {
        val events = mutableListOf<IntArray>()
        for (b in buildings) {
            events.add(intArrayOf(b[0], -b[2]))   // start: negative height
            events.add(intArrayOf(b[1], b[2]))    // end:   positive height
        }
        events.sortWith(Comparator { a, b -> if (a[0] != b[0]) a[0] - b[0] else a[1] - b[1] })

        // max-heap of active heights; seed with 0 = ground level
        val pq = PriorityQueue<Int>(compareByDescending { it })
        pq.offer(0)

        val result = mutableListOf<List<Int>>()
        var prevMax = 0
        for (e in events) {
            if (e[1] < 0) pq.offer(-e[1])   // building start
            else pq.remove(e[1])            // building end (O(n))
            val currMax = pq.peek()
            if (currMax != prevMax) {
                result.add(listOf(e[0], currMax))
                prevMax = currMax
            }
        }
        return result
    }
}
